Add controllers

alanpodolsk opened the pull request.

alanpodolsk committed:
1

alanpodolsk committed:
Users CRUD done

alanpodolsk committed:
Items CRUD done

alanpodolsk committed:
Fixed SpotBug Errors

kesch9 commented on src/main/java/ru/practicum/shareit/booking/BookingDto.java:
```diff
@@ -0,0 +1,27 @@
+package ru.practicum.shareit.booking;
```
Зря вы перенесли модели и dto в один пакет, в макете проекта была верная структура, обычно dto и model не смешивают и у вас возможно(в зависимости от вашей реализации) будут добавлять еще dto, лучше вернуть структуру как была

kesch9 replied:
Это для всех сущностей

kesch9 commented on src/main/java/ru/practicum/shareit/exception/ErrorHandler.java:
```diff
@@ -0,0 +1,36 @@
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.CONFLICT)
+    public Map<String, String> handleConflictException(final ConflictException e) {
+        return Map.of("Conflict error", e.getMessage());
```
Можно сделать более верно и отправлять объект (dto), а не маппу 
Например ErrorResponce у которого будет поле String error куда передадите сообщение ошибки

alanpodolsk replied:
Исправлено

kesch9 commented on src/main/java/ru/practicum/shareit/item/Item.java:
```diff
@@ -0,0 +1,22 @@
+
+/**
+ * TODO Sprint add-controllers.
+ */
```
Todo можно удалить, где подправили код

kesch9 replied:
Не подправлено

kesch9 commented on src/main/java/ru/practicum/shareit/item/Item.java:
```diff
@@ -0,0 +1,22 @@
+@AllArgsConstructor
+@NoArgsConstructor
+public class Item {
+    Integer id;
```
Не забываем проставлять private для полей

kesch9 replied:
Не подправлено

kesch9 commented on src/main/java/ru/practicum/shareit/item/ItemDto.java:
```diff
@@ -0,0 +1,30 @@
+    Boolean available;
+    Integer requestId;
+
+    public static class ItemMapper {
```
В Dto не хранятся мапперы, они не для этого. Нужно вынести мапперы в отдельный класс, также не хватает методов dto->model

alanpodolsk replied:
Исправлено

kesch9 replied:
Mappers не хранят в dto, пакет dto для хранения dto

kesch9 commented on src/main/java/ru/practicum/shareit/item/ItemRepository.java:
```diff
@@ -0,0 +1,17 @@
+import java.util.Map;
+
+@Component
+public class ItemRepository {
```
Repository нужно добавить CRUD операции, а не делать это в сервисных методах, дальше вы поменяете репозитории который храниться в памяти на БД и там эти операции будут в самом репозитории

alanpodolsk replied:
Исправлено

kesch9 commented on src/main/java/ru/practicum/shareit/item/ItemRepository.java:
```diff
@@ -0,0 +1,17 @@
+
+@Component
+public class ItemRepository {
+    static Map<Integer, Item> items = new HashMap<>();
```
static удаляем, добавляем private final

kesch9 commented on src/main/java/ru/practicum/shareit/item/ItemServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+        checkItem(item);
+        item.setId(itemRepository.getNewId());
+        item.setOwner(userService.getUser(userId));
+        itemRepository.items.put(item.getId(), item);
```
Все операции CRUD внутри репозиториев

kesch9 commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,42 @@
+    ItemService itemService;
+
+    @PostMapping
+    public Item createItem(@RequestBody Item item, @RequestHeader("X-Sharer-User-Id") Integer userId) {
```
На вход и на выход методов в контроллере нужно использовать DTO, они для этого и нужны, в сервисе также это подправить, а вот в репозиториях уже модели используются

kesch9 commented on src/main/java/ru/practicum/shareit/item/ItemServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+@Component
+@AllArgsConstructor
+public class ItemServiceImpl implements ItemService {
+    private UserService userService;
```
Здесь лучше использовать сам репозиторий, чтобы получать именно модели

kesch9 commented on src/main/java/ru/practicum/shareit/item/ItemDto.java:
```diff
@@ -0,0 +1,30 @@
+@AllArgsConstructor
+@NoArgsConstructor
+public class ItemDto {
+    String name;
```
private добавляем, попробуйте также провалидировать через аннотации lombok, для этого например помечаем поле name @NotBlank, но это поле надо валидировать только при создании создаем интерфейс Create
public interface Create {
}
и добавляем в аннотацию
@NotBlank(groups = {Create.class})
А в методе контроллера делаем вот так
public ItemDto createItem(@RequestHeader("X-Sharer-User-Id") long userId, @RequestBody @Validated(Create.class) ItemDto item) 
Теперь поле будет валидировать по ломбок, так удобнее, попробуйте переделать. Дальше вам это пригодиться

kesch9 commented on src/main/java/ru/practicum/shareit/item/ItemServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+        if (text.isBlank()) {
+            return items;
+        }
+        for (Item item : itemRepository.items.values()) {
```
попробуйте переделать через stream используя filter

kesch9 commented on src/main/java/ru/practicum/shareit/user/User.java:
```diff
@@ -1,7 +1,17 @@
+@AllArgsConstructor
+@NoArgsConstructor
 public class User {
+    Integer id;
```
Замечания такие же как и в item дублируются, дальше буду писать специфичные ошибки для User

kesch9 commented on src/main/java/ru/practicum/shareit/user/User.java:
```diff
@@ -1,7 +1,17 @@
 public class User {
+    Integer id;
+    String name;
+    String email;
```
Для валидации email есть аннотация ломбок @Email

kesch9 commented on src/main/java/ru/practicum/shareit/user/UserRepository.java:
```diff
@@ -0,0 +1,20 @@
+public class UserRepository {
+    static Map<Integer, User> users = new HashMap<>();
+    private static Integer currentId = 0;
+    static List<String> uniqueEmails = new ArrayList<>();
```
Удаляем static и делаем private final

kesch9 reviewed (commented):
Нужно доработать работу

alanpodolsk committed:
Доработки согласно замечаниям:
1. Валидация по аннотациям;
2. Использование Dto в контроллерах и сервисах
3. Перенос CRUD-операций в репозитории
4. Минорные изменения

kesch9 commented on src/main/java/ru/practicum/shareit/item/ItemRepository.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    public List<Item> getItemsByUser(User user) {
+        List userItems = new ArrayList<>();
```
Нужно указывать тип хранящих данных
List<Item>

kesch9 commented on src/main/java/ru/practicum/shareit/item/dto/ItemMapper.java:
```diff
@@ -0,0 +1,28 @@
+
+import ru.practicum.shareit.item.model.Item;
+
+public class ItemMapper {
```
Для утильных классов есть аннотация ломбок @UtilityClass

kesch9 commented on src/main/java/ru/practicum/shareit/user/UserRepository.java:
```diff
@@ -0,0 +1,54 @@
+public class UserRepository {
+    private final Map<Integer, User> users = new HashMap<>();
+    private Integer currentId = 0;
+    private final Map<Integer,String> uniqueEmails = new HashMap<>();
```
Для уникальных email можно использовать Set, он здесь лучше подходит, чем map

kesch9 commented on src/main/java/ru/practicum/shareit/user/UserRepository.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public void emailIsDuplicate(String email) {
+        if (uniqueEmails.values().contains(email)) {
```
Можно писать вот так
uniqueEmails.containsValue(email)

kesch9 commented on src/main/java/ru/practicum/shareit/item/ItemServiceImpl.java:
```diff
@@ -0,0 +1,112 @@
+        for (Item item : items) {
+            itemDtos.add(ItemMapper.toItemDto(item));
+        }
+        return itemDtos;
```
Не подправили через stream приведу пример как можно переделать
return items.stream().map(ItemMapper::toItemDto).collect(Collectors.toList());
надо учиться использовать stream сейчас пишут только через них, как вы написали уже считается устаревшим

kesch9 reviewed (approved):
Approve
